### DISCUSSION TITLE
Installing a python.extension_module into a venv

### DISCUSSION THREAD

**Discussion by 8uurg** [mesonbuild/meson#10113] [Q&A] [answered]
When using `meson install -C build`, an extension_module marked with `install: true` will be installed for the global system installation, even when called from a virtualenv. From meson's source code there should exist an option `python.install_env`, which could be set to `venv` or `auto` to accomplish this. This option is however not configurable.

Are there any other alternatives for installing an extension module into a virtualenv?

**Comment by eli-schwartz** [comment, marked as answer]
I added this option in the Meson 0.62.0 development cycle, which means it does not exist *yet* in released versions of Meson.

You can install and test out yesterday's 0.62.0rc1 preview and see if it works for you. :)

**Reply by 8uurg** [reply]
Thank you! Using the RC setting `python.install_env` to `auto` or `venv` indeed works. Thanks for implementing this, did not expect that this was a recent addition!

**Reply by adam2392** [reply]
Hi, I tried adding this to my `meson.build` file:

```
# Setup Python:
# https://mesonbuild.com/Python-module.html
py3_mod = import('python')

# NOTE: with Meson >=0.64.0 we can add `pure: false` here and remove that line
# everywhere else, see https://github.com/mesonbuild/meson/pull/10783.
py3 = py3_mod.find_installation(
  'python3',
  pure: false  # Will be installed next to binaries
)
message(py3.full_path())
message(py3.get_install_dir())
py3.install_env('venv')

if py3.language_version().version_compare('<3.8')
    error('At least Python 3.8 is required.')
endif
py3_dep = py3.dependency()
```

and got the following error:

```
>>> meson build --wipe
The Meson build system
Version: 1.0.0
Source dir: /Users/adam2392/Documents/scikit-tree
Build dir: /Users/adam2392/Documents/scikit-tree/build
Build type: native build
Project name: scikit-tree
Project version: 0.0.0.dev0
C compiler for the host machine: arm64-apple-darwin20.0.0-clang (clang 14.0.6 "clang version 14.0.6")
C linker for the host machine: arm64-apple-darwin20.0.0-clang ld64 609
C++ compiler for the host machine: arm64-apple-darwin20.0.0-clang++ (clang 14.0.6 "clang version 14.0.6")
C++ linker for the host machine: arm64-apple-darwin20.0.0-clang++ ld64 609
Host machine cpu family: aarch64
Host machine cpu: aarch64
Library m found: YES
Program cython found: YES (/Users/adam2392/miniforge3/envs/sktree/bin/cython)
Program python3 found: YES (/Users/adam2392/miniforge3/envs/sktree/bin/python3)
Message: /Users/adam2392/miniforge3/envs/sktree/bin/python3
Message: /opt/homebrew/lib/python3.9/site-packages/

meson.build:60:4: ERROR: Unknown method "install_env" in object <[PythonInstallation] holds [PythonExternalProgram]: <PythonExternalProgram 'python3' -> ['/Users/adam2392/miniforge3/envs/sktree/bin/python3']>> of type PythonInstallation.

A full log can be found at /Users/adam2392/Documents/scikit-tree/build/meson-logs/meson-log.txt
WARNING: Running the setup command as `meson [options]` instead of `meson setup [options]` is ambiguous and deprecated.
```

Is this the correct usage? I could not find a relevant example in the meson docs

**Reply by eli-schwartz** [reply]
https://mesonbuild.com/Builtin-options.html#python-module

**Reply by kaimast** [reply]
This seems to be broken now: `../meson.build:51:4: ERROR: Unknown method "install_env" in object <[PythonInstallation] holds [PythonExternalProgram]: <PythonExternalProgram 'python3' -> ['/usr/bin/python']>> of type PythonInstallation.`

**Reply by eli-schwartz** [reply]
You're having the same issue as the post two above yours -- the post immediately above yours was a reference to the `meson setup` command line arguments indicating that what you are trying isn't supposed to work.

Meson doesn't allow you to conditionally set command line options dynamically in your build scripts. The value selected at configure time (via the command line or via a default value) is finalized and used everywhere.

Note that users might not want to install into a virtual environment so permitting it to be unconditionally overridden via a function would lead to problems anyway.

Pass -D options to `meson setup` instead. You can also set the default value in `project(..., default_options: ['option=valuexxx'])` which still allows users to set that option to something else if they really want.

**Reply by kaimast** [reply]
Oh, sorry. I figured this out right before you replied. Thank you for replying so quickly!!

I still get `../meson.build:49:26: ERROR: python.install_env cannot be set to "venv" unless you are in a venv!` even though I am in a virtual environment.

**Reply by eli-schwartz** [reply]
Probably the detected Python program somehow ended up being the Python from out of the venv, e.g. it might have picked up the one that meson was installed with instead. You can enforce the correct python via a native file and meson-python does this automatically for you.

**Reply by kaimast** [reply]
meson-python seems to work. Thanks again for your help :)
